Zend/zend_inheritance.c

```diff
@@ -1053,9 +1055,19 @@ static void ZEND_COLD emit_incompatible_method_error(
 			}
 		}
 	} else {
-		zend_error_at(E_COMPILE_ERROR, func_filename(child), func_lineno(child),
-			"Declaration of %s must be compatible with %s",
-			ZSTR_VAL(child_prototype), ZSTR_VAL(parent_prototype));
+		if (child->op_array.scope->num_traits) {
+			zend_error_at(E_COMPILE_ERROR, func_filename(child), func_lineno(child),
+				"Declaration of %s (from %s) must be compatible with %s",
+				ZSTR_VAL(child_prototype), ZSTR_VAL(child->op_array.scope->trait_names[0].name), ZSTR_VAL(parent_prototype));
```

Comment:
This looks wrong, as a function doesn't necessarily come from a trait, and not always from the first trait. AFAIK we don't track from what trait a method originated from. We can probably find out by looking at the class/parent class/traits, i.e. check some trait we use declares the method, or we rename the method with `as` to the given name. Additionally, we need to make sure the method actually comes from that trait, i.e. we're not using `insteadof`. The class itself must also nott overshadow the method. This check could be tricky, as at this point in time you're looking at the class with traits already inherited.

I might be wrong, maybe there is some flag to discriminate.

Reply:
That's right. They were already integrated. I can't really find a way to figure out from which trait the method came from... because we don't check it. As I said, this is a preliminary idea :-)